rfcs/20200601-tfx-udsl-semantics.md

```diff
@@ -0,0 +1,427 @@
+2.  Adapt the pipeline scheduling to the pipeline running time. In the context
+    of the previous example, the pipeline will be scheduled to run every day.
+3.  Split the pipeline into multiple parts, each of which employs a different
+    schedule.
+
+Option 1 is fine in an ideal world, where the computation resource is unlimited.
+However in real world, this option will likely cause pipeline runs to pile up in
+a certain stage which is problematic due to resource contention. Option 2 will
+not have the resource contention problem but it gives up the benefit of the
+timely-arriving data. It is likely that the `Trainer` is training on some old
+data despite of more recent data being available, which is a compromise of model
+freshness and is not ideal for scenarios that are sensitive to model freshness.
+Option 3 can potentially solve the problem but is not user-friendly and is prone
+to future changes of individual components and pipeline shape.
+
+On the other hand, if we are able to support components with different schedules
+in a pipeline, the problem can be naturally solved.
+
+### Synchronization barrier
+
+It is not recommended to have concurrent pipeline runs since it will cause
+problems for components that need to be guarded by a 'synchronization barrier'
+such as `Pusher`. However, this is not avoidable sometimes even if the intention
+is not to have pipeline runs concurrently.
+
+<div style="text-align:center"><img src='20200601-tfx-udsl-semantics/daily_job_normal.png', width='550', alt='dj_normal', title='Healthy daily jobs'></div>
+
+The figure above shows a daily ML pipeline in an ideal world. For
+simplification, we combine all components but `Pusher` into a single stage. As
+you can see, we expect a pipeline run to finish within a day so it will not have
+overlap with other runs. However, there might be cases that the pipeline
+overruns and result in concurrent pipeline runs. The figure below shows an
+extreme but possible case, where the Tuesday run actually finishes after the
+Wednesday run. This is problematic since Tuesday run will push an older model
+into production which is likely to cause regression.
+
+<div style="text-align:center"><img src='20200601-tfx-udsl-semantics/daily_job_abnormal.png', width='600', alt='dj_abnormal', title='Problematic daily jobs'></div>
+
+To address the problem, we need to guarantee that:
+
+*   There is only one instance of `Pusher` running at any time.
+*   Each `Pusher` run always pushes a model that is considered better than any
+    previously pushed model.
+
+However, there is no good solution to provide such guarantees without compromise
+in the existing TFX DSL.
+
+## Design proposal
+
+### A new way to categorize ML pipelines
+
+Before going into the design details, we would like to introduce a new way to
+categorize ML pipelines. Understanding these concepts is critical to evaluate
+the rest of this RFC.
+
+#### Synchronous execution & Asynchronous execution
```

Comment:
It's not totally clear to me how this will work with orchestrators like Airflow and KubeFlow. Do these natively support asynchronous execution? Or is the idea that Synchronous pipelines should use an orchestrator, and Async pipelines are managed as a cluster?

Reply:
Thanks for the question @cweill ! There are multiple ways to run an Async execution pipeline:

- A pub-sub system will work
- A centralized 'orchestrator' / 'controller' will also work
- A collection of long-running jobs in a cluster will also work

The goal of this RFC is to propose the semantics and does not intend to propose / recommend an implementation pattern.

> It's not totally clear to me how this will work with orchestrators like Airflow and KubeFlow

These two platforms are good for synchronous execution pipelines but might not be the best platforms to run asynchronous execution pipelines. As the two patterns (sync vs async) are quite different, we expect that people use different platforms for them.

Hope this answers your question :)

Reply:
The most simple way to run asynchronous execution pipelines will be just deploying it to a cluster through a controller. No specific platform like Airflow or Kubeflow is needed.
Pub-sub systems are natural fits for this style of pipelines on the other hand. Beam / Dataflow is definitely another option using its support for unbounded source.

> Also wondering, it this may be solved w a unified TFX UI that does support those (just thinking aloud)

A unified UI is definitely something we always keep in mind and is in our list. @zhitaoli

Reply:
Is the idea to provide some sort of orchestrator/runner for this on the long run? (aka a custom controller that can be deployed on GCP for async pipelines?)

Reply:
Btw, kinda related to this topic. Airflow community is talking to introduce a similar concept (which would be great as that would mean that Airflow could implement the Async pipeline execution). Thought, I would bring it up here. Curious on your thoughts on this. Note the current state is dicussion so it's still under the process of deciding what the implementation will look like.  https://lists.apache.org/thread.html/re1a7e5cfcb1e9f4a0bfac41998da2d88ffb26d4f597036c772a4c86e%40%3Cdev.airflow.apache.org%3E

Reply:
@ruoyu90 Can we forward this RFC to the Airflow community and seek for comment too? I imagine there is a possibility to implement the async style semantic on top of the future Airflow implementation in the future.

Reply:
> The most simple way to run asynchronous execution pipelines will be just deploying it to a cluster through a controller. 

Do you mean deploying the pipeline components as a collection of standalone always running nodes?

Reply:
@paveldournov yes exactly.

/cc @hongye-sun